thunder/core/jit_ext.py

```diff
@@ -773,6 +773,125 @@ def backward_impl(*args, **kwargs):
+@register_general_jit_lookaside(torch.ops.higher_order.autograd_function_apply)
+def _general_jit_torch_ops_higher_order_autograd_function_apply(fwd, bwd, *fwd_args, **fwd_kwargs):
+    from thunder.core import utils
+    from thunder.core.baseutils import sequencify
+    from thunder.core.pytree import tree_flatten, tree_map
+    from thunder.core.transforms import VJPDual, augmented_forward_impls, backward_impls
+
+    def _generate_random_str_id() -> str:
+        import secrets
+        import string
+
+        length = 5
+        return "".join(secrets.choice(string.ascii_lowercase) for _ in range(length))
+
+    jit_ctx: JitCtx = get_jit_ctx()
+
+    args_tensor_mask = unwrap(fwd_kwargs["args_tensor_mask"])
+    # TODO(crcrpar): Think about making use of `non_differentiable_idx`
+    # note that this key is quite new: https://github.com/pytorch/pytorch/pull/134087
+    # non_differentiable_idx = fwd_kwargs.get("non_differentiable_idx")
+    length_of_tensor_args = sum(args_tensor_mask)
+    new_fwd_args = (wrap_const(None),) + fwd_args[:length_of_tensor_args]
+    old_scope = jit_ctx.computation_trace.scopes
+    fwd_bsyms = []
+    jit_ctx.computation_trace.scopes = [fwd_bsyms]
+
+    fwd_result = _interpret_call(fwd, *new_fwd_args)
+    if fwd_result is INTERPRETER_SIGNALS.EXCEPTION_RAISED:
+        return fwd_result
+    output, saved_values = unwrap(fwd_result)
+    wrapped_output = wrap(output, provenance=fwd_result.provenance)
+
+    unwrapped_fwd_args = tree_map(lambda t: unwrap(t), new_fwd_args)[1:]
+
+    producer_map = utils.producers(fwd_bsyms)
+    tensor_to_prod_bsym: dict[Variable, BoundSymbol] = {}
+    for p in tree_flatten((output, saved_values))[0]:
+        if not isinstance(p, TensorProxy):
+            continue
+        if p in producer_map:
+            prod_bsym = producer_map[p]
+            tensor_to_prod_bsym[variableify(p)] = prod_bsym
+    prod_bsym_to_tensor = {v: unvariableify(k) for k, v in tensor_to_prod_bsym.items()}
+
+    # Encapsulate custom fwd into a bsym.
+    sym_id = f"autograd_function_apply_{_generate_random_str_id()}"
+    sym = Symbol(
+        name=sym_id,
+        id=sym_id,
+        _module=fwd_bsyms[-1].sym.module,
+    )
+    bsym_of_custom_autograd_func = BoundSymbol(
+        sym,
+        args=unwrapped_fwd_args,
+        kwargs={},
+        output=output,
+        subsymbols=fwd_bsyms,
+        header=(
+            f"output of fwd_body: {output}, saved_values from fwd_body: "
+            f"{[t.name if isinstance(t, Proxy) else t for t in saved_values]}"
+        ),
+        source_filename=jit_ctx.computation_trace._current_source_filename,
+        source_positions=None,
+        _call_ctx=fwd_bsyms[0]._call_ctx,
+        _import_ctx=fwd_bsyms[0]._import_ctx,
+        _object_ctx=fwd_bsyms[0]._object_ctx,
+        _executor=fwd_bsyms[0]._executor,
+    )
+    old_scope[-1].append(bsym_of_custom_autograd_func)
+
+    # Define augmented fwd rule and backward rule on the fly.
```

Comment:
Once you have transformed forward and backward functions to be Thunder friendly can you call `thunder.torch.autograd_function_apply` from here instead of doing on-the-fly registration?